lbry/lbry/conf.py

```diff
@@ -362,6 +362,9 @@ def __delitem__(self, key):
         del self.data[key]
 
 
+TBC = typing.TypeVar('TBC', bound='BaseConfig')
```

Comment:
What's this for?

Reply:
this is how the internet told me to do typing for a @classmethod

Reply:
doesn't seem right, especially for a `@classmethod`, the `cls` variable points to the class on which the class method appears (it's bound to it)

I think you'd only need to do this if you wanted to pass the Config to a function somewhere else, but you didn't want to bind it to `BaseConfig` or to the specific sublcass of `BaseConfig`. I think everywhere in LBRY it should require the specific `Config` object that describes LBRY configs, so you can just use that directly without needing a `TypeVar`.

see here for correct usage:
https://www.python.org/dev/peps/pep-0484/#type-variables-with-an-upper-bound

Reply:
a) how can I say the function returns a `Config` object when I'm in the `BaseConfig` class? I don't really understand that part.
b) if everything uses Config and not BaseConfig, why have both classes? can we merge the two?

btw here's the link I got that from: https://github.com/python/typing/issues/58#issuecomment-326240794. It's endorsed by gvanrossum :wink:

Reply:
I got the opposite impression from what Guido says in that thread:
"We're not going to support this. The issue seems fairly academic."

a) I think for the "Factory" pattern specifically, you will need the `TypeVar` after all, but in the PR you have it marked on the `cls` variable too, that's what I was mostly objecting to. The `cls` variable already has the correct type info, if you hover over it in PyCharm it should show correct type.

b) Makes it easier to test in isolation and also if we wanted to have another Config type, for example, if we wanted to re-use this code in the wallet server (which currently has its own crappy config management implementation).

So, in summary, I would only have the `TBC` set as return type of `def create_from_arguments(cls, args) -> TBC:` and not on the `cls` variables since those are typed correctly already (also, you can't call those class methods and pass the `cls` as an argument, the `@classmethod` takes care of that, adding the type info to `cls` is just noise and doesn't affect the API or internal operation of `@classmethod`).